### PR TITLE
8262470: Printed GlyphVector outline with low DPI has bad quality on Windows

### PR DESCRIPTION
**Pull request by AlexanderScherbatiy**
Printing text using GlyphVector outline has bad quality on printers with low DPI on Windows.
The GDI system used for text printing on Windows accepts only integer path coordinates.
Rounding GlyphVector outline coordinates leads to distorted printed text.

The issue had been reported as JDK-8256264 but was reverted because of the regression JDK-8259007 "This test printed a blank page".

The fix JDK-8256264 scaled coordinates in wPrinterJob.moveTo()/lineTo()  methods up and scaled transforms in wPrinterJob.beginPath()/endPath() down.

The regression was in the WPathGraphics.deviceDrawLine() method which uses wPrinterJob.moveTo()/lineTo() methods without surrounding them with wPrinterJob.beginPath()/endPath() so the line coordinates were only scaled up.

I tried to put wPrinterJob.beginPath()/endPath()  methods around wPrinterJob.moveTo()/lineTo()   in the method WPathGraphics.deviceDrawLine()  but the line was not drawn at all even without scaling coordinates up and transform down (without JDK-8256264 fix). It looks like GDI treats this case as an empty shape.

The proposed fix applies path coordinates and transform scaling only in WPathGraphics.convertToWPath() method.
The one more PathPrecisionScaleFactorShapeTest.java manual test is added which checks that all methods that draw paths in WPathGraphics are used: line in WPathGraphics.deviceDrawLine() and SEG_LINETO/SEG_QUADTO/SEG_CUBICTO in WPathGraphics.convertToWPath() .

The `java/awt/print` and `java/awt/PrintJob` automatic and manual tests were run on Windows 10 Pro with the fix.

There are two failed automated tests which fail without the fix as well:
```
java/awt/print/PrinterJob/GlyphPositions.java 
java/awt/print/PrinterJob/PSQuestionMark.java
```

The following manual tests have issues on my system:
- `java/awt/print/Dialog/PrintDlgPageable.java` java.lang.IllegalAccessException: class com.sun.javatest.regtest.agent.MainWrapper$MainThread cannot access a member of class PrintDlgPageable with modifiers "public static"

- `java/awt/print/PrinterJob/PrintAttributeUpdateTest.java` I select pages radio button, press the print button but the test does not finish and I do not see any other dialogs with pass/fail buttons.

- `java/awt/PrintJob/PrintCheckboxTest/PrintCheckboxManualTest.java` Tests that there is no ClassCastException thrown in printing checkbox and scrollbar with XAWT. Error. Can't find HTML file: test\jdk\java\awt\PrintJob\PrintCheckboxTest\PrintCheckboxManualTest.html

- `java/awt/print/PrinterJob/SecurityDialogTest.java` A windows with instructions is shown but it does not contain  print/pass/fail buttons and it is not possible to close the window.

- The tests below fail with "Error. Parse Exception: Arguments to `manual' option not supported: yesno" message:
```
java/awt/print/Dialog/DialogOrient.java
java/awt/print/Dialog/DialogType.java
java/awt/print/PrinterJob/ImagePrinting/ClippedImages.java
java/awt/print/PrinterJob/ImagePrinting/ImageTypes.java
java/awt/print/PrinterJob/ImagePrinting/PrintARGBImage.java
java/awt/print/PrinterJob/PageDialogTest.java
java/awt/print/PrinterJob/PageRanges.java
java/awt/print/PrinterJob/PageRangesDlgTest.java
java/awt/print/PrinterJob/PrintGlyphVectorTest.java
java/awt/print/PrinterJob/PrintLatinCJKTest.java
java/awt/print/PrinterJob/PrintTextTest.java
java/awt/print/PrinterJob/SwingUIText.java
java/awt/PrintJob/ConstrainedPrintingTest/ConstrainedPrintingTest.java
java/awt/PrintJob/PageSetupDlgBlockingTest/PageSetupDlgBlockingTest.java
java/awt/PrintJob/SaveDialogTitleTest.java
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262470](https://bugs.openjdk.java.net/browse/JDK-8262470): Printed GlyphVector outline with low DPI has bad quality on Windows

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2756/head:pull/2756` \
`$ git checkout pull/2756`

Update a local copy of the PR: \
`$ git checkout pull/2756` \
`$ git pull https://git.openjdk.java.net/jdk pull/2756/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2756`

View PR using the GUI difftool: \
`$ git pr show -t 2756`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2756.diff">https://git.openjdk.java.net/jdk/pull/2756.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
I'll run the tests for this fix.

**Review by mrserb — APPROVED** [review]

**Comment by mrserb** [comment]
@prsadhuk please take a look

File: src/java.desktop/windows/classes/sun/awt/windows/WPrinterJob.java [hunk 1]
```diff
@@ -995,6 +1009,44 @@ protected void setPolyFillMode(int fillRule) {
...
+        return setAdvancedGraphicsMode(getPrintDC());
+    }
+
+    /**
+     * Set the GDI graphics mode.
+     * The {@code mode} should
+     * be one of the following Windows constants:
+     * {@code GM_COMPATIBLE} or {@code GM_ADVANCED}.
+     */
+    private int setGraphicsMode(int mode) {
```
Comment (prsadhuk) [inline src/java.desktop/windows/classes/sun/awt/windows/WPrinterJob.java:1025]: Is there any need of "int" return value? I dont see it is used in restoreTransform()

File: src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp [hunk 2]
```diff
@@ -1932,6 +1932,111 @@ JNIEXPORT void JNICALL Java_sun_awt_windows_WPrinterJob_setPolyFillMode
...
+    elems = env->GetDoubleArrayElements(transform, 0);
+
+    xForm.eM11 = (FLOAT) elems[0];
+    xForm.eM12 = (FLOAT) elems[1];
+    xForm.eM21 = (FLOAT) elems[2];
+    xForm.eM22 = (FLOAT) elems[3];
+    xForm.eDx  = (FLOAT) elems[4];
+    xForm.eDy  = (FLOAT) elems[5];
+
+    ::SetWorldTransform((HDC)printDC, &xForm);
```
Comment (prsadhuk) [inline src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp:2033]: Probably we should check for the return value of all this system APIs SetGraphicsMode, GetWorldTransform, SetWorldTransform, ModifyWorldTransform to see if it succeeded?

File: src/java.desktop/windows/classes/sun/awt/windows/WPrinterJob.java (hunk 1 above)
Comment (AlexanderScherbatiy) [reply src/java.desktop/windows/classes/sun/awt/windows/WPrinterJob.java]: I updated the code to return void from setGraphicsMode() method.

File: src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp (hunk 2 above)
Comment (AlexanderScherbatiy) [reply src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp]: I added DASSERT to check SetGraphicsMode and Get/Set/ModifyWorldTransform results.

**Review by prsadhuk — APPROVED** [review]

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**